Visual C++ Code Analysis broken by commit 99c425a

Microsoft Visual C++ 2019 version 16.3.3, following simple module:

```cpp
#include "wil/com.h"

int main() {}
```

Code Analysis works fine with f74a30b64a48a4df13f34fcd824a2f8c108b7748
```
1>------ Build started: Project: wil_analyze, Configuration: Debug Win32 ------
1>wil_analyze.cpp
========== Build: 1 succeeded, 0 failed, 0 up-to-date, 0 skipped ==========
```

but fails with 99c425a12c0e48da7e7b172c448827b34496c4a9 https://github.com/microsoft/wil/pull/80
```
1>------ Build started: Project: wil_analyze, Configuration: Debug Win32 ------
1>wil_analyze.cpp
1>E:\Source\wil_analyze\wil\include\wil\result_macros.h(3617,1): error C2059: syntax error: '<L_GRAMMAR_template_head_l>'
1>E:\Source\wil_analyze\wil\include\wil\result_macros.h(3618,1): error C2143: syntax error: missing ';' before '{'
1>E:\Source\wil_analyze\wil\include\wil\result_macros.h(3618,1): error C2447: '{': missing function header (old-style formal list?)
1>E:\Source\wil_analyze\wil\include\wil\result_macros.h(3626,1): error C2059: syntax error: ''template<''
1>E:\Source\wil_analyze\wil\include\wil\result_macros.h(3628,1): error C2143: syntax error: missing ';' before '{'
...
1>E:\Source\wil_analyze\wil\include\wil\result_macros.h(5301,31): error C2039: 'ReportFailure_NtStatus': is not a member of 'wil::details'
1>E:\Source\wil_analyze\wil\include\wil\result_macros.h(5301,53): fatal error C1003: error count exceeds 100; stopping compilation
1>Done building project "wil_analyze.vcxproj" -- FAILED.
========== Build: 0 succeeded, 1 failed, 0 up-to-date, 0 skipped ==========

```

I wondered why Code Analysis works fine in https://github.com/microsoft/terminal - but WIL is held back there on working version e8c599bca6c56c44b6730ad93f6abbc9ecd60fc1

--- a/include/wil/result_macros.h
+++ b/include/wil/result_macros.h
@@ -3611,9 +3611,9 @@ __WI_SUPPRESS_4127_E
             }
         }
 
+        template<FailureType T>
         _Success_(true)
         _Translates_Win32_to_HRESULT_(err)
-        template<FailureType T>
         __declspec(noinline) inline HRESULT ReportFailure_Win32(__R_FN_PARAMS_FULL, DWORD err)
         {
             const auto hr = __HRESULT_FROM_WIN32(err);
@@ -3621,18 +3621,18 @@ __WI_SUPPRESS_4127_E
             return hr;
         }
         
+        template<>
         _Success_(true)
         _Translates_Win32_to_HRESULT_(err)
-        template<>
         __declspec(noinline) inline RESULT_NORETURN HRESULT ReportFailure_Win32<FailureType::FailFast>(__R_FN_PARAMS_FULL, DWORD err)
         {
             const auto hr = __HRESULT_FROM_WIN32(err);
             ReportFailure<FailureType::FailFast>(__R_FN_CALL_FULL, hr);
         }
-         
+
+        template<>
         _Success_(true)
         _Translates_Win32_to_HRESULT_(err)
-        template<>
         __declspec(noinline) inline RESULT_NORETURN HRESULT ReportFailure_Win32<FailureType::Exception>(__R_FN_PARAMS_FULL, DWORD err)
         {
             const auto hr = __HRESULT_FROM_WIN32(err);
@@ -3664,9 +3664,9 @@ __WI_SUPPRESS_4127_E
             ReportFailure<FailureType::Exception>(__R_FN_CALL_FULL, hr);
         }
 
+        template<FailureType T>
         _Success_(true)
         _Translates_last_error_to_HRESULT_
-        template<FailureType T>
         __declspec(noinline) inline HRESULT ReportFailure_GetLastErrorHr(__R_FN_PARAMS_FULL)
         {
             const auto hr = GetLastErrorFailHr(__R_FN_CALL_FULL);
@@ -3674,27 +3674,27 @@ __WI_SUPPRESS_4127_E
             return hr;
         }
 
+        template<>
         _Success_(true)
         _Translates_last_error_to_HRESULT_
-        template<>
         __declspec(noinline) inline RESULT_NORETURN HRESULT ReportFailure_GetLastErrorHr<FailureType::FailFast>(__R_FN_PARAMS_FULL)
         {
             const auto hr = GetLastErrorFailHr(__R_FN_CALL_FULL);
             ReportFailure<FailureType::FailFast>(__R_FN_CALL_FULL, hr);
         }
 
+        template<>
         _Success_(true)
         _Translates_last_error_to_HRESULT_
-        template<>
         __declspec(noinline) inline RESULT_NORETURN HRESULT ReportFailure_GetLastErrorHr<FailureType::Exception>(__R_FN_PARAMS_FULL)
         {
             const auto hr = GetLastErrorFailHr(__R_FN_CALL_FULL);
             ReportFailure<FailureType::Exception>(__R_FN_CALL_FULL, hr);
         }
 
+        template<FailureType T>
         _Success_(true)
         _Translates_NTSTATUS_to_HRESULT_(status)
-        template<FailureType T>
         __declspec(noinline) inline HRESULT ReportFailure_NtStatus(__R_FN_PARAMS_FULL, NTSTATUS status)
         {
             const auto hr = wil::details::NtStatusToHr(status);
@@ -3702,18 +3702,18 @@ __WI_SUPPRESS_4127_E
             return hr;
         }
 
+        template<>
         _Success_(true)
         _Translates_NTSTATUS_to_HRESULT_(status)
-        template<>
         __declspec(noinline) inline RESULT_NORETURN HRESULT ReportFailure_NtStatus<FailureType::FailFast>(__R_FN_PARAMS_FULL, NTSTATUS status)
         {
             const auto hr = wil::details::NtStatusToHr(status);
             ReportFailure<FailureType::FailFast>(__R_FN_CALL_FULL, hr);
         }
 
+        template<>
         _Success_(true)
         _Translates_NTSTATUS_to_HRESULT_(status)
-        template<>
         __declspec(noinline) inline RESULT_NORETURN HRESULT ReportFailure_NtStatus<FailureType::Exception>(__R_FN_PARAMS_FULL, NTSTATUS status)
         {
             const auto hr = wil::details::NtStatusToHr(status);
@@ -3762,9 +3762,9 @@ __WI_SUPPRESS_4127_E
             ReportFailure_Msg<FailureType::Exception>(__R_FN_CALL_FULL, hr, formatString, argList);
         }
 
+        template<FailureType T>
         _Success_(true)
         _Translates_Win32_to_HRESULT_(err)
-        template<FailureType T>
         __declspec(noinline) inline HRESULT ReportFailure_Win32Msg(__R_FN_PARAMS_FULL, DWORD err, _Printf_format_string_ PCSTR formatString, va_list argList)
         {
             auto hr = __HRESULT_FROM_WIN32(err);
@@ -3772,18 +3772,18 @@ __WI_SUPPRESS_4127_E
             return hr;
         }
 
+        template<>
         _Success_(true)
         _Translates_Win32_to_HRESULT_(err)
-        template<>
         __declspec(noinline) inline RESULT_NORETURN HRESULT ReportFailure_Win32Msg<FailureType::FailFast>(__R_FN_PARAMS_FULL, DWORD err, _Printf_format_string_ PCSTR formatString, va_list argList)
         {
             auto hr = __HRESULT_FROM_WIN32(err);
             ReportFailure_Msg<FailureType::FailFast>(__R_FN_CALL_FULL, hr, formatString, argList);
         }
 
+        template<>
         _Success_(true)
         _Translates_Win32_to_HRESULT_(err)
-        template<>
         __declspec(noinline) inline RESULT_NORETURN HRESULT ReportFailure_Win32Msg<FailureType::Exception>(__R_FN_PARAMS_FULL, DWORD err, _Printf_format_string_ PCSTR formatString, va_list argList)
         {
             auto hr = __HRESULT_FROM_WIN32(err);
@@ -3815,10 +3815,9 @@ __WI_SUPPRESS_4127_E
             ReportFailure_Msg<FailureType::Exception>(__R_FN_CALL_FULL, hr, formatString, argList);
         }
 
-        _Success_(true)
-        _When_((type == FailureType::Exception) || (type == FailureType::FailFast), _Analysis_noreturn_)
-        _Translates_last_error_to_HRESULT_
         template<FailureType T>
+        _Success_(true)
+        _Translates_last_error_to_HRESULT_
         __declspec(noinline) inline HRESULT ReportFailure_GetLastErrorHrMsg(__R_FN_PARAMS_FULL, _Printf_format_string_ PCSTR formatString, va_list argList)
         {
             auto hr = GetLastErrorFailHr(__R_FN_CALL_FULL);
@@ -3826,27 +3825,27 @@ __WI_SUPPRESS_4127_E
             return hr;
         }
 
+        template<>
         _Success_(true)
         _Translates_last_error_to_HRESULT_
-        template<>
         __declspec(noinline) inline RESULT_NORETURN HRESULT ReportFailure_GetLastErrorHrMsg<FailureType::FailFast>(__R_FN_PARAMS_FULL, _Printf_format_string_ PCSTR formatString, va_list argList)
         {
             auto hr = GetLastErrorFailHr(__R_FN_CALL_FULL);
             ReportFailure_Msg<FailureType::FailFast>(__R_FN_CALL_FULL, hr, formatString, argList);
         }
 
+        template<>
         _Success_(true)
         _Translates_last_error_to_HRESULT_
-        template<>
         __declspec(noinline) inline RESULT_NORETURN HRESULT ReportFailure_GetLastErrorHrMsg<FailureType::Exception>(__R_FN_PARAMS_FULL, _Printf_format_string_ PCSTR formatString, va_list argList)
         {
             auto hr = GetLastErrorFailHr(__R_FN_CALL_FULL);
             ReportFailure_Msg<FailureType::Exception>(__R_FN_CALL_FULL, hr, formatString, argList);
         }
 
+        template<FailureType T>
         _Success_(true)
         _Translates_NTSTATUS_to_HRESULT_(status)
-        template<FailureType T>
         __declspec(noinline) inline HRESULT ReportFailure_NtStatusMsg(__R_FN_PARAMS_FULL, NTSTATUS status, _Printf_format_string_ PCSTR formatString, va_list argList)
         {
             auto hr = wil::details::NtStatusToHr(status);
@@ -3854,18 +3853,18 @@ __WI_SUPPRESS_4127_E
             return hr;
         }
 
+        template<>
         _Success_(true)
         _Translates_NTSTATUS_to_HRESULT_(status)
-        template<>
         __declspec(noinline) inline RESULT_NORETURN HRESULT ReportFailure_NtStatusMsg<FailureType::FailFast>(__R_FN_PARAMS_FULL, NTSTATUS status, _Printf_format_string_ PCSTR formatString, va_list argList)
         {
             auto hr = wil::details::NtStatusToHr(status);
             ReportFailure_Msg<FailureType::FailFast>(__R_FN_CALL_FULL, hr, formatString, argList);
         }
 
+        template<>
         _Success_(true)
         _Translates_NTSTATUS_to_HRESULT_(status)
-        template<>
         __declspec(noinline) inline RESULT_NORETURN HRESULT ReportFailure_NtStatusMsg<FailureType::Exception>(__R_FN_PARAMS_FULL, NTSTATUS status, _Printf_format_string_ PCSTR formatString, va_list argList)
         {
             auto hr = wil::details::NtStatusToHr(status);
@@ -4068,6 +4067,7 @@ __WI_SUPPRESS_4127_E
             // Log Macros
             //*****************************************************************************
 
+            _Post_satisfies_(return == hr)
             __R_DIRECT_METHOD(HRESULT, Log_Hr)(__R_DIRECT_FN_PARAMS HRESULT hr) WI_NOEXCEPT
             {
                 __R_FN_LOCALS;
@@ -4322,7 +4322,8 @@ __WI_SUPPRESS_4127_E
                 return status;
             }
 
-            __R_DIRECT_METHOD(HRESULT, Log_HrMsg)(__R_DIRECT_FN_PARAMS HRESULT hr, PCSTR formatString, ...) WI_NOEXCEPT
+            _Post_satisfies_(return == hr)
+            __R_DIRECT_METHOD(HRESULT, Log_HrMsg)(__R_DIRECT_FN_PARAMS HRESULT hr, _Printf_format_string_ PCSTR formatString, ...) WI_NOEXCEPT
             {
                 va_list argList;
                 va_start(argList, formatString);
@@ -4331,7 +4332,8 @@ __WI_SUPPRESS_4127_E
                 return hr;
             }
 
-            __R_DIRECT_METHOD(DWORD, Log_Win32Msg)(__R_DIRECT_FN_PARAMS DWORD err, PCSTR formatString, ...) WI_NOEXCEPT
+            _Post_satisfies_(return == err)
+            __R_DIRECT_METHOD(DWORD, Log_Win32Msg)(__R_DIRECT_FN_PARAMS DWORD err, _Printf_format_string_ PCSTR formatString, ...) WI_NOEXCEPT
             {
                 va_list argList;
                 va_start(argList, formatString);
@@ -4729,6 +4731,7 @@ __WI_SUPPRESS_4127_E
                 }
             }
 
+            _Post_satisfies_(return == condition) _When_(condition, _Analysis_noreturn_)
             __RFF_CONDITIONAL_METHOD(bool, FailFast_HrIf)(__RFF_CONDITIONAL_FN_PARAMS HRESULT hr, bool condition) WI_NOEXCEPT
             {
                 if (condition)
@@ -5094,6 +5097,7 @@ __WI_SUPPRESS_4127_E
                 wil::details::ReportFailure_Hr<FailureType::FailFast>(__RFF_INTERNAL_FN_CALL E_UNEXPECTED);
             }
 
+            _Post_satisfies_(return == condition) _When_(condition, _Analysis_noreturn_)
             __RFF_CONDITIONAL_METHOD(bool, FailFast_If)(__RFF_CONDITIONAL_FN_PARAMS bool condition) WI_NOEXCEPT
             {
                 if (condition)
@@ -5103,6 +5107,7 @@ __WI_SUPPRESS_4127_E
                 return condition;
             }
 
+            _Post_satisfies_(return == condition) _When_(!condition, _Analysis_noreturn_)
             __RFF_CONDITIONAL_METHOD(bool, FailFast_IfFalse)(__RFF_CONDITIONAL_FN_PARAMS bool condition) WI_NOEXCEPT
             {
                 if (!condition)
@@ -5220,6 +5225,7 @@ __WI_SUPPRESS_4127_E
                 return hr;
             }
 
+            _Post_satisfies_(return == condition) _When_(condition, _Analysis_noreturn_)
             __RFF_CONDITIONAL_METHOD(bool, FailFastImmediate_If)(bool condition) WI_NOEXCEPT
             {
                 if (condition)

